src/v/wasm/allocator.cc

```diff
@@ -26,36 +31,99 @@
+    if (_size < req.minimum || _size > req.maximum) {
         return std::nullopt;
     }
     heap_memory front = std::move(_memory_pool.front());
     _memory_pool.pop_front();
     return front;
 }
 
+ss::future<> heap_allocator::wait_for_available_memory() {
+    // Out of an abundance of caution, add a timeout to this wait, zeroing
+    // should not take this long to execute even for the largest chunks of
+    // memory.
+    constexpr static auto timeout = std::chrono::seconds(1);
+    // Wait for the pool to have memory available or for no async zeroing tasks
+    // to be in flight.
+    return _cond_var.wait(ss::lowres_clock::now() + timeout, [this] {
+        return !_memory_pool.empty() || _gate.get_count() == 0;
+    });
+}
+
 void heap_allocator::deallocate(heap_memory m, size_t used_amount) {
-    std::memset(m.data.get(), 0, used_amount);
+    if (used_amount <= _memset_chunk_size) {
+        std::memset(m.data.get(), 0, used_amount);
+        _memory_pool.push_back(std::move(m));
+        _cond_var.signal();
+        return;
+    }
+    // We don't use ssx::with_gate because we need the gate to be destroyed in
+    // the same continuation as the condition_variable notifications, see the
+    // comment in async_deallocate for more details.
+    ssx::background = async_deallocate(_gate.hold(), std::move(m), used_amount);
+}
+
+ss::future<> heap_allocator::async_deallocate(
+  ss::gate::holder _, heap_memory m, size_t used_amount) {
+    {
+        uint8_t* data = m.data.get();
+        size_t remaining = used_amount;
+        while (true) {
+            if (remaining <= _memset_chunk_size) {
+                std::memset(data, 0, remaining);
+                break;
+            }
+            std::memset(data, 0, _memset_chunk_size);
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+            data += _memset_chunk_size;
+            remaining -= _memset_chunk_size;
+            co_await ss::coroutine::maybe_yield();
+        }
+    }
     _memory_pool.push_back(std::move(m));
+    // If we're the only one holding the gate open, then broadcast all the
+    // waiters to wake up. This prevents waiters from getting stuck when no
+    // memory is being deallocated.
+
+    // Scenario when this can happen:
+    // - an async deallocation starts
+    // - **two** waiters come
+    // - the dealloc finishes, and we need to notify both waiters
+    //   - tell one that the pool is no longer empty
+    //   - tell the other that the there are no zeroing tasks and memory
+    //     allocation is going to fail.
+
+    // Now you might be asking yourself why the gate count will be zero when the
+    // gate held open by this task. The key thing here is that the broadcast
+    // checks the condition asynchronously, so this continuation will complete
+    // and the gate released before the condition in `wait_for_available_memory`
+    // is checked.
+    if (_gate.get_count() == 1) {
```

Comment:
Alternatively, we could keep track of the number of waiters, and ensure that we don't have more waiters than the number of inflight deallocations. Thoughts?

Reply:
i guess my #1 question is if broadcast/signal here is an optimization, or related to correctness. if the later, then it seems a bit fishy. maybe ok, but maybe there is a more direct way to do this instead of needing to mention anything about the internals of the condvar broadcast method.

>  Now you might be asking yourself why the gate count will be zero when the gate held open by this task.

i wasn't asking myself that (because we are holding the gate here as you point out), but we also aren't checking if it is zero (just 1 or non-1). so maybe i'm missing something here?

>Alternatively, we could keep track of the number of waiters, and ensure that we don't have more waiters than the number of inflight deallocations. Thoughts?

could you replace the gate+condvar with a semaphore (or gate + semaphore if the gate is useful for its traditional use case) and then check the semaphore waiters? no real preference here, just throwing ideas out that might be a more direct representation of what's happening.

Reply:
Okay simplified this a ton by localizing the hacking around wasmtime's API via a thread local. PTAL, hopefully the semantics are much clearer now